core/src/main/java/org/apache/calcite/util/ArrowSet.java

```diff
@@ -0,0 +1,357 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.ArrayDeque;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents a set of functional dependencies.
+ * Each functional dependency is represented as an {@link Arrow} from a set of
+ * determinant columns to a set of dependent columns.
+ *
+ * <p>An {@link ArrowSet} captures the complete set of functional dependencies
+ * that hold in a relation. Provides core algorithms in functional dependency theory.
+ *
+ * @see Arrow
+ * @see ImmutableBitSet
+ */
+public class ArrowSet {
+  public static final ArrowSet EMPTY = new ArrowSet(ImmutableSet.of());
+
+  // Maps each determinant set to the dependent set it functionally determines.
+  private final Map<ImmutableBitSet, ImmutableBitSet> dependencyGraph = new HashMap<>();
+
+  // Maps each column ordinal to its containing determinant sets for efficient reverse lookup.
+  private final Map<Integer, Set<ImmutableBitSet>> reverseIndex = new HashMap<>();
```

Comment:
"its containing" is not clear. 
Do you mean "the keys of the dependencyGraph which contain this column?"